cmake/AddStandaloneTests.cmake

```diff
@@ -0,0 +1,122 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+# Helper function to set up a CMake target for a test executable.  It
+# can safely be called multiple times for the same executable and will
+# use the LINK_LIBRARIES passed on the first call.
+function(add_standalone_test_executable EXECUTABLE_NAME LINK_LIBRARIES)
+  if (TARGET ${EXECUTABLE_NAME})
+    return()
+  endif()
+
+  add_spectre_executable(
+    ${EXECUTABLE_NAME}
+    ${EXECUTABLE_NAME}.cpp
+    )
+
+  add_dependencies(
+    ${EXECUTABLE_NAME}
+    module_GlobalCache
+    module_Main
+    )
+
+  target_link_libraries(
+    ${EXECUTABLE_NAME}
+    PRIVATE
+    ${LINK_LIBRARIES}
+    )
+
+  add_dependencies(test-executables ${EXECUTABLE_NAME})
+endfunction()
+
+# For tests that result in a failure it is necessary to redirect
+# output from stderr to stdout. However, it was necessary at least on
+# some systems to do this redirect inside a shell command.
+find_program(SHELL_EXECUTABLE "sh")
+if (NOT SHELL_EXECUTABLE)
+  message(FATAL_ERROR
+    "Could not find 'sh' shell to execute standalone failure tests")
+endif()
+
+# Add a standalone test named TEST_NAME that runs an executable with
+# no arguments.  A test named Foo.Bar.Baz will run the executable
+# Test_Baz.  If REGEX_TO_MATCH is nonempty the test will pass if the
+# output matches it, otherwise the test will pass if the executable
+# succeeds without any "ERROR" output.
+function(add_standalone_test TEST_NAME LINK_LIBRARIES REGEX_TO_MATCH)
+  string(REGEX MATCH "[^.]*$" EXECUTABLE_NAME "${TEST_NAME}")
+  set(EXECUTABLE_NAME "Test_${EXECUTABLE_NAME}")
+  add_standalone_test_executable("${EXECUTABLE_NAME}" "${LINK_LIBRARIES}")
+  add_test(
+    NAME "${TEST_NAME}"
+    COMMAND
+    ${SHELL_EXECUTABLE}
+    -c
+    "${CMAKE_BINARY_DIR}/bin/${EXECUTABLE_NAME} 2>&1"
+    )
+
+  if("${REGEX_TO_MATCH}" STREQUAL "")
+    set_tests_properties(
+      "${TEST_NAME}"
+      PROPERTIES
+      FAIL_REGULAR_EXPRESSION "ERROR"
+      TIMEOUT 10
+      LABELS "standalone"
+      ENVIRONMENT "ASAN_OPTIONS=detect_leaks=0")
+  else()
+    set_tests_properties(
+      "${TEST_NAME}"
+      PROPERTIES
+      PASS_REGULAR_EXPRESSION
+      "${REGEX_TO_MATCH}"
+      TIMEOUT 10
+      LABELS "standalone"
+      ENVIRONMENT "ASAN_OPTIONS=detect_leaks=0")
+  endif()
+endfunction()
+
+# Add a standalone test named TEST_NAME that runs an executable
+# passing a --input-file command-line option.  A test named
+# Foo.Bar.Baz will by default run the executable Test_Baz with the
+# input file Test_Baz.yaml, but an optional third argument can
+# override the executable name to Test_THIRD_ARG.  A fourth argument,
```

Comment:
Maybe have the argument completely override the executable name, without prepending `Test_`? I think that would be easier to understand when calling this function, because at that point the user has probably already created an executable with this name.

Reply:
Enforcing that the execs begin with `Test_` sounds good. My suggestion would be to assert that the `EXECUTABLE_NAME` has the prefix, rather than prepending it, but I leave this up to you. If you keep the prepending-logic, please make that clear in the documentation of the `EXECUTABLE` argument of the `add_standalone_test` function above.